contrib/format-msgpack/src/main/java/org/apache/drill/exec/store/msgpack/MsgpackSchema.java

```diff
@@ -0,0 +1,114 @@
+import org.apache.hadoop.security.AccessControlException;
+
+import com.google.protobuf.TextFormat;
+import com.google.protobuf.TextFormat.ParseException;
+
+public class MsgpackSchema {
+  public static final String SCHEMA_FILE_NAME = ".schema.proto";
+
+  @SuppressWarnings("unused")
+  private static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(MsgpackSchema.class);
+
+  private DrillFileSystem fileSystem;
+
+  public MsgpackSchema(DrillFileSystem fileSystem) {
+    this.fileSystem = fileSystem;
+  }
+
+  public MaterializedField load(Path schemaLocation) throws AccessControlException, FileNotFoundException, IOException {
+    MaterializedField previousMapField = null;
+    if (schemaLocation != null && fileSystem.exists(schemaLocation)) {
+      try (FSDataInputStream in = fileSystem.open(schemaLocation)) {
+        String schemaData = IOUtils.toString(in);
+        Builder newBuilder = SerializedField.newBuilder();
+        try {
+          TextFormat.merge(schemaData, newBuilder);
+        } catch (ParseException e) {
+          throw new DrillRuntimeException("Failed to read schema file: " + schemaLocation, e);
+        }
+        SerializedField read = newBuilder.build();
+        previousMapField = MaterializedField.create(read);
+      }
+    }
+    return previousMapField;
+  }
+
+  public void save(MaterializedField mapField, Path schemaLocation) throws IOException {
+    try (FSDataOutputStream out = fileSystem.create(schemaLocation, true)) {
+      SerializedField serializedMapField = mapField.getSerializedField();
+      String data = TextFormat.printToString(serializedMapField);
+      IOUtils.write(data, out);
+    }
+  }
+
+  public MaterializedField merge(MaterializedField existingField, MaterializedField newField) {
+    Preconditions.checkArgument(existingField.getType().getMinorType() == MinorType.MAP,
+        "Field " + existingField + " is not a MAP type.");
+    Preconditions.checkArgument(existingField.hasSameTypeAndMode(newField),
+        "Field " + existingField + " and " + newField + " not same.");
+    MaterializedField merged = existingField.clone();
+    privateMerge(merged, newField);
+    return merged;
+  }
+
+  private void privateMerge(MaterializedField existingField, MaterializedField newField) {
+    Preconditions.checkArgument(existingField.getType().getMinorType() == MinorType.MAP,
+        "Field " + existingField + " is not a MAP type.");
+    for (MaterializedField newChild : newField.getChildren()) {
+      String newChildName = newChild.getName();
+      MaterializedField foundExistingChild = getFieldByName(newChildName, existingField);
+      if (foundExistingChild != null) {
+        if (foundExistingChild.hasSameTypeAndMode(newChild)) {
+          if (foundExistingChild.getType().getMinorType() == MinorType.MAP) {
+            privateMerge(foundExistingChild, newChild);
+          } // else we already have it
+        } else {
+          // error

```

Comment:
MsgPack items are probably all nullable (as in JSON). Does it support only two three as in JSON: Number, String and Boolean?

The challenge here is choosing a numeric format (BIGINT or DOUBLE), and knowing how to convert from one to the other.

It is somewhat difficult to extract the rules being used just from the code. Maybe a Javadoc comment or package-info file to give some background?

Reply:
The fields are all nullable, correct. It supports many "sizes" for Number but in the java library I'm using it rolls all those up into Number. So yes there's just a few types to handle. But there is a distinction between Number and Float. So it's not as difficult as in JSON where you might see "1", "2" then "3.1".

Again I will add more comments and documentation.